woocommerce/woocommerce-gateway-stripe
# Add SCA prompt if required prior to "Change payment method" form submission

# Changes proposed in this Pull Request:

Fixes https://github.com/woocommerce/woocommerce-gateway-stripe/issues/1046

Branches off of https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1398

The main change (https://github.com/woocommerce/woocommerce-gateway-stripe/commit/38733a56779239e268cc7ef7f2a06b038cc342b8) is only a few lines of code, to expand the SCA prompt for the "Add payment method" form (in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1398) to the "Change payment method" form for subscriptions as well. (The preceding two commits are refactors to the relevant code that are not necessary for or directly related to the main fix – happy to drop them if that's better for any reason.)

This approach entirely supercedes https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1405, which redirects to an SCA step after saving the changed payment method – this is less desirable than authenticating prior to form submission. The reason I'd initially gone with approach in #1405 was to keep it similar to the "Pay for order" form, which seemed like a good idea at the time because it is so similar to the "Change payment method" form: the latter uses the same route, clears the former from the output buffer, and replaces it with a similar form (with the same ID).

However, with regard to SCA, it is more similar to the "Add payment method" form in terms of setting up the payment method outside the context of a payment (using a SetupIntent) and so there may not be any reason to follow the precedent of calling `process_payment` prior to the SCA challenge. Therefore, the exact same approach is borrowed, which addresses all of the remaining concerns in #1405. Assuming it's correct that there is no benefit to delaying the SCA prompt until after the form is submitted, the only caveat is just that this implementation has to look for a hidden input inside the form (to distinguish this form from the "Pay for order" form rather than checking the identifier (which is the kind of somewhat more brittle condition I'd originally hoped to avoid).

After attempting form submission with a card that requires SCA, the prompt appears over the payment form (here using [regulatory test card](https://stripe.com/docs/testing#regulatory-cards) in test mode):

<img width="896" src="https://user-images.githubusercontent.com/1867547/102748459-3749b280-4330-11eb-805c-87d69f806f44.png">

After failure:

<img width="816" src="https://user-images.githubusercontent.com/1867547/102749058-53018880-4331-11eb-9f51-70ebd385a3d9.png">

After success:

<img width="1119" src="https://user-images.githubusercontent.com/1867547/102749054-5268f200-4331-11eb-91f0-9f63f200443d.png">

# Testing instructions

From #1405:

> Use instructions from https://github.com/woocommerce/woocommerce-gateway-stripe/issues/1046 but **verify that modal authentication appears**, and that if it succeeds then **the renewal also succeeds**. (Just note that the "Edit Subscription" screen needs to be reloaded after the payment method is saved, or else the "Process renewal" action will actually update it back to the earlier source, being the value of the "Stripe Source ID" field in the admin form 🤔.)

> (To change the payment method, navigate on the front-end to My account » Subscriptions » any active subscription and click "Change payment", or navigate in wp-admin to WooCommerce » Subscriptions » any active subscription and click "Customer change payment method page →".)

> Can also test to make sure the ordinary Pay for Order flow (for regular pending orders) still works as before.

-------------------
